substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/OldGeneration.java

```diff
@@ -42,70 +51,213 @@
+            fixingVisitor.setChunk(aChunk);
+            RelocationInfo.walkObjects(aChunk, fixingVisitor);
+            aChunk = HeapChunk.getNext(aChunk);
+        }
+        timers.tenuredFixingAlignedChunks.close();
+
+        timers.tenuredFixingImageHeap.open();
+        HeapImpl.getHeapImpl().walkImageHeapObjects(fixingVisitor);
+        timers.tenuredFixingImageHeap.close();
+
+        timers.tenuredFixingThreadLocal.open();
+        ThreadLocalMTWalker.walk(refFixingVisitor);
+        timers.tenuredFixingThreadLocal.close();
+
+        timers.tenuredFixingRuntimeCodeCache.open();
+        refFixingVisitor.debug = true;
+        if (RuntimeCompilation.isEnabled()) {
+            RuntimeCodeInfoMemory.singleton().walkRuntimeMethodsDuringGC(runtimeCodeCacheWalker);
+        }
+        refFixingVisitor.debug = false;
+        timers.tenuredFixingRuntimeCodeCache.close();
+
+        timers.tenuredFixingStack.open();
+        GCImpl.getGCImpl().blackenStackRoots(refFixingVisitor);
+        timers.tenuredFixingStack.close();
+
+        timers.tenuredFixingUnalignedChunks.open();
+        UnalignedHeapChunk.UnalignedHeader uChunk = space.getFirstUnalignedHeapChunk();
+        while (uChunk.isNonNull()) {
+            Log.log().string("[OldGeneration.fixing: fixing phase, chunk=").zhex(uChunk)
+                    .string(", unaligned]").newline().flush();
+            Pointer objPointer = UnalignedHeapChunk.getObjectStart(uChunk);
+            Object obj = objPointer.toObject();
+            fixingVisitor.visitObject(obj);
+            uChunk = HeapChunk.getNext(uChunk);
+        }
+        timers.tenuredFixingUnalignedChunks.close();
+    }
+
+    void compacting(Timers timers) {
+        // Phase 3: Copy objects to their new location
+        AlignedHeapChunk.AlignedHeader chunk = space.getFirstAlignedHeapChunk();
+        while (chunk.isNonNull()) {
+            Log trace = Log.noopLog().string("[OldGeneration.compacting: chunk=").zhex(chunk);
+
+            if (chunk.getFirstRelocationInfo().isNull()) {
+                /*
+                 * No compaction necessary as there are no gaps.
+                 */
+                trace.string(", skip");
+            } else if (false /* TODO */) {
+                /*
+                 * Skip compaction as fragmentation isn't severe enough.
+                 */
+                trace.string(", skip");
+            } else {
+                trace.string(", firstRelocationInfo=").zhex(chunk.getFirstRelocationInfo());
+                trace.string(", oldTop=").zhex(HeapChunk.getTopPointer(chunk));
+
+                timers.tenuredCompactingCunks.open();
+                compactingVisitor.init(chunk);
+                RelocationInfo.walkObjects(chunk, compactingVisitor);
+                compactingVisitor.finish();
+                timers.tenuredCompactingCunks.close();
+
+                timers.tenuredUpdatingRememberedSet.open();
+                RememberedSet.get().clearRememberedSet(chunk);
```

Comment:
I think you don't need to clean the set first, that happens as part of `enableRememberedSetForChunk` below, too.